### PR TITLE
feat[lang]: support top level `"abi"` key in json interfaces

### PR DESCRIPTION
**Pull request by sandbubbles**
### What I did
address https://github.com/vyperlang/vyper/issues/4231

### Commit message

```
some frameworks produce json ABI files with `"abi"` as a top-level
key. this decreases a bit of friction for using those ABI files.
```
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/unit/cli/vyper_json/test_compile_json.py [hunk 1]
```diff
@@ -295,3 +295,27 @@ def test_relative_import_paths(input_json):
...
+      ],
+      "outputs": [{ "name": "max_stream_life", "type": "uint256" }]
+    }
+  ]
+}
+    """
+    code = """
+from . import json
+    """
+    input_bundle = make_input_bundle({"json.json": json, "code.vy": code})
```
Comment (charles-cooper) [inline tests/unit/cli/vyper_json/test_compile_json.py:320]: maybe let's call the file `abi.json` or `stream.json`

**Review by charles-cooper — COMMENTED** [review]
looks good. did you find the commit where the regression was introduced?

**Comment by sandbubbles** [comment]
> looks good. did you find the commit where the regression was introduced?

No no, should i look into it?

**Comment by charles-cooper** [comment]
> > looks good. did you find the commit where the regression was introduced?
> 
> No no, should i look into it?

yes please -- i was going to include it when drafting the commit message

**Comment by sandbubbles** [comment]
> > > looks good. did you find the commit where the regression was introduced?
> > 
> > 
> > No no, should i look into it?
> 
> yes please -- i was going to include it when drafting the commit message

I think it was always there.
```
6b8b7f6c558cff79f5b552ed8a13df296b469a58 is the first bad commit
commit 6b8b7f6c558cff79f5b552ed8a13df296b469a58
Author: Ben Hauser <ben@hauser.id>
Date:   Fri Aug 21 17:15:06 2020 +0300

    refactor: allow access to vyper-json via vyper-compile

 vyper/cli/vyper_compile.py | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)
 ```
 
 when run against 
 ```
 {
  "abi": [
    {
      "name": "validate",
      "inputs": [
        { "name": "creator", "type": "address" },
        { "name": "token", "type": "address" },
        { "name": "amount_per_second", "type": "uint256" },
        { "name": "reason", "type": "bytes" }
      ],
      "outputs": [{ "name": "max_stream_life", "type": "uint256" }]
    }
  ]
}
```
The error was the same across all commits.

**Comment by charles-cooper** [comment]
something is weird, because we theoretically used to have code to handle this, e.g. https://github.com/vyperlang/vyper/blob/9136169468f317a53b4e7448389aa315f90b95ba/vyper/cli/vyper_compile.py#L264-L268

**Comment by cyberthirst** [comment]
> something is weird, because we theoretically used to have code to handle this, e.g.
> 
> https://github.com/vyperlang/vyper/blob/9136169468f317a53b4e7448389aa315f90b95ba/vyper/cli/vyper_compile.py#L264-L268

i didn't look deeply into the code, but in this link we don't flatten the "abi" dict into the list (which is what we do in this fix)

**Comment by charles-cooper** [comment]
> > something is weird, because we theoretically used to have code to handle this, e.g.
> > https://github.com/vyperlang/vyper/blob/9136169468f317a53b4e7448389aa315f90b95ba/vyper/cli/vyper_compile.py#L264-L268
> 
> i didn't look deeply into the code, but in this link we don't flatten the "abi" dict into the list (which is what we do in this fix)

so why is this considered a regression :thinking: paging @fubuloubu

**Comment by fubuloubu** [comment]
The regression might be that it didn't try and check anything was implemented because it was empty?

This fix seems good regardless

**Review by fubuloubu — APPROVED** [review]

**Review by pcaversaccio — APPROVED** [review]
For Solidity development frameworks, this can be quite common. For example, using Hardhat, you get [this](https://github.com/pcaversaccio/createx/blob/main/artifacts/src/CreateX.sol/CreateX.json#L5) as an example.

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**